pkg/controller/populators/clone-populator.go

```diff
@@ -0,0 +1,368 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/controller/clone"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+)
+
+const (
+	// LabelOwnedByUID label sets the relationship between a resource and "owner"
+	LabelOwnedByUID = "cdi.kubevirt.io/OwnedByUID"
+
+	// AnnClonePhase tracks the status of the clone op
+	AnnClonePhase = "cdi.kubevirt.io/clonePhase"
+
+	// AnnCloneError has the error string for error phase
+	AnnCloneError = "cdi.kubevirt.io/cloneError"
+
+	clonePopulatorName = "clone-populator"
+
+	cloneFinalizer = "cdi.kubevirt.io/clonePopulator"
+
+	pendingPhase = "Pending"
+
+	succeededPhase = "Succeeded"
+
+	errorPhase = "Error"
+)
+
+// Planner is an interface to mock out planner implementation for testing
+type Planner interface {
+	ChooseStrategy(context.Context, *clone.ChooseStrategyArgs) (*cdiv1.CDICloneStrategy, error)
+	Plan(context.Context, *clone.PlanArgs) ([]clone.Phase, error)
+}
+
+// ClonePopulatorReconciler reconciles PVCs with VolumeCloneSources
+type ClonePopulatorReconciler struct {
+	ReconcilerBase
+	planner Planner
+}
+
+// NewClonePopulator creates a new instance of the clone-populator controller
+func NewClonePopulator(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	clonerImage string,
+	pullPolicy string,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &ClonePopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(clonePopulatorName),
+			recorder:        mgr.GetEventRecorderFor(clonePopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			sourceKind:      cdiv1.VolumeCloneSourceRef,
+			installerLabels: installerLabels,
+		},
+	}
+
+	clonePopulator, err := controller.New(clonePopulatorName, mgr, controller.Options{
+		MaxConcurrentReconciles: 5,
+		Reconciler:              reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	planner := &clone.Planner{
+		RootObjectType:  &corev1.PersistentVolumeClaimList{},
+		OwnershipLabel:  LabelOwnedByUID,
+		UIDField:        uidField,
+		Image:           clonerImage,
+		PullPolicy:      corev1.PullPolicy(pullPolicy),
+		InstallerLabels: installerLabels,
+		Client:          reconciler.client,
+		Recorder:        reconciler.recorder,
+		Controller:      clonePopulator,
+	}
+	reconciler.planner = planner
+
+	if err := addCommonPopulatorsWatches(mgr, clonePopulator, log, cdiv1.VolumeCloneSourceRef, &cdiv1.VolumeCloneSource{}); err != nil {
```

Comment:
Maybe we need to add a watch for clones without source, so we retrigger the reconcile once the source PVC is created.

Reply:
Currently will requeue if the planner returns nil.  This covers the source not existing case.  https://github.com/mhenriks/containerized-data-importer/blob/clone-pop/pkg/controller/populators/clone-populator.go#L207

There is a functional test for this case.

Removing the requeue with a watch is doable but it is a kind of complicated.  Will require an index on the source field of VolumeCloneSource